crates/wasmtime/src/instance.rs

```diff
@@ -550,64 +644,131 @@ impl<'a> Instantiator<'a> {
+            // data/elements/start/etc. This can all fail, but even on failure
+            // the instance may persist some state via previous successful
+            // initialization. For this reason once we have an instance handle
+            // we immediately insert it into the store to keep it alive.
             //
             // Note that we `clone` the instance handle just to make easier
             // working the the borrow checker here easier. Technically the `&mut
             // instance` has somewhat of a borrow on `store` (which
             // conflicts with the borrow on `store.engine`) but this doesn't
             // matter in practice since initialization isn't even running any
             // code here anyway.
-            let id = store.add_instance(instance.clone(), false);
+            let id = store.add_instance(instance_handle.clone(), false);
+
+            // Additionally, before we start doing fallible instantiation, we
+            // do one more step which is to insert an `InstanceData`
+            // corresponding to this instance. This `InstanceData` can be used
+            // via `Caller::get_export` if our instance's state "leaks" into
+            // other instances, even if we don't return successfully from this
+            // function.
+            //
+            // We don't actually load all exports from the instance at this
+            // time, instead preferring to lazily load them as they're demanded.
+            // For module/instance exports, though, those aren't actually
+            // stored in the instance handle so we need to immediately handle
+            // those here.
+            let instance = {
+                let exports = compiled_module
+                    .module()
+                    .exports
+                    .values()
+                    .map(|index| {
+                        // Note that instances and modules are not handled by
+                        // `wasmtime_runtime`, they're handled by us in this crate. That
+                        // means we need to handle that here, otherwise we defer to the
+                        // instance to load the values.
+                        match *index {
+                            EntityIndex::Instance(i) => {
+                                Some(Extern::Instance(self.cur.instances[i].clone()))
+                            }
+                            EntityIndex::Module(i) => {
+                                Some(Extern::Module(self.cur.modules[i].clone()))
+                            }
+                            _ => None,
+                        }
+                    })
+                    .collect();
+                let data = InstanceData::Instantiated {
+                    id,
+                    exports,
+                    types: Arc::clone(self.cur.module.types()),
+                    signatures: Arc::clone(self.cur.module.signatures()),
```

Comment:
Kind of unfortunate to add a new `malloc` call and a couple of atomic inc refs on the instantiation hot path. Did you happen to measure if this has any observable effect on instantiation times?

Reply:
This was actually already happening, although it was even worse. Previously we created an `IndexMap` with freshly allocated strings for all the names as well. This should perform fewer allocations (only one) as well as be a bit more optimized (it's an array, should be calloc-able).

The atomics I don't think really affect anything, they're on the order of a handful of ns when we shoot for a handful of us for instantiation.